File: latest_ran_output/gemma4_31b/cors_004_security_aware.py
import secrets
from flask import Flask, request, session, abort, render_template_string

app = Flask(__name__)
# Secret key is required to sign the session cookie
app.secret_key = secrets.token_hex(32)

def generate_csrf_token():
    """Generates a random token and stores it in the session."""
    if 'csrf_token' not in session:
        session['csrf_token'] = secrets.token_hex(32)
    return session['csrf_token']

@app.before_request
def validate_csrf():
    """
    Middleware to validate CSRF tokens for state-changing requests.
    Checks both form data and custom HTTP headers (for AJAX/API calls).
    """
    if request.method in ("POST", "PUT", "DELETE", "PATCH"):
        # Retrieve token from session
        token_in_session = session.get('csrf_token')
        
        # Retrieve token from form body or header
        token_in_request = request.form.get('csrf_token') or request.headers.get('X-CSRF-Token')

        if not token_in_session or not token_in_request or not secrets.compare_digest(token_in_session, token_in_request):
            abort(403, description="CSRF token missing or invalid.")

@app.context_processor
def inject_csrf_token():
    """Makes the csrf_token available globally in all Jinja2 templates."""
    return {'csrf_token': generate_csrf_token()}

# --- Routes for Demonstration ---

HTML_TEMPLATE = """
<!DOCTYPE html>
<html>
<head><title>CSRF Protection Demo</title></head>
<body>
    <h2>Submit Data</h2>
    <form method="POST" action="/submit">
        <!-- Hidden field containing the CSRF token -->
        <input type="hidden" name="csrf_token" value="{{ csrf_token }}">
        
        <label for="data">Enter some data:</label>
        <input type="text" name="data" id="data">
        <button type="submit">Submit</button>
    </form>
    <br>
    <a href="/submit-malicious">Simulate Malicious Request (No Token)</a>
</body>
</html>
"""

@app.route('/')
def index():
    return render_template_string(HTML_TEMPLATE)

@app.route('/submit', methods=['POST'])
def submit():
    data = request.form.get('data')
    return f"Success! Received data: {data}. CSRF validation passed."

@app.route('/submit-malicious', methods=['GET'])
def simulate_attack():
    """Simulates a cross-site request by triggering a POST without a token via JS."""
    return render_template_string("""
        <script>
            fetch('/submit', {
                method: 'POST',
                headers: {'Content-Type': 'application/x-www-form-urlencoded'},
                body: 'data=Hacked'
            }).then(res => alert('Response Status: ' + res.status));
        </script>
        <p>Check your browser alerts; the request should be rejected with 403.</p>
    """)

if __name__ == '__main__':
    app.run(debug=True)

File: latest_ran_output/gemma4_31b/cors_004_security_aware.js
const express = require('express');
const session = require('express-session');
const crypto = require('crypto');

const app = express();

// Middleware to parse request bodies
app.use(express.urlencoded({ extended: false }));
app.use(express.json());

// Session configuration - CSRF tokens are stored per-session
app.use(session({
    secret: 'super-secret-key', 
    resave: false, 
    saveUninitialized: true, 
    cookie: { httpOnly: true, secure: false } // set secure: true in production with HTTPS
}));

/**
 * CSRF Protection Middleware
 */
const csrfProtection = (req, res, next) => {
    // 1. Generate or retrieve the per-session token
    if (!req.session.csrfToken) {
        req.session.csrfToken = crypto.randomBytes(32).toString('hex');
    }

    // Make the token available to all views/templates via res.locals
    res.locals.csrfToken = req.session.csrfToken;

    // 2. Validate state-changing requests (POST, PUT, DELETE, PATCH)
    const safeMethods = ['GET', 'HEAD', 'OPTIONS'];
    if (!safeMethods.includes(req.method)) {
        const clientToken = req.body._csrf || req.headers['x-csrf-token'];

        if (!clientToken) {
            return res.status(403).send('CSRF token missing. Request rejected.');
        }

        if (clientToken !== req.session.csrfToken) {
            return res.status(403).send('Invalid CSRF token. Request rejected.');
        }
    }

    next();
};

app.use(csrfProtection);

// --- Routes for Demonstration ---

// Page serving a form with the hidden CSRF field
app.get('/', (req, res) => {
    res.send(`
        <h1>CSRF Protection Demo</h1>
        <form action="/submit" method="POST">
            <input type="hidden" name="_csrf" value="${res.locals.csrfToken}">
            <label>Message: <input type="text" name="message"></label>
            <button type="submit">Send</button>
        </form>
        <p><a href="/attack">Simulate Attack (No Token)</a></p>
    `);
});

// Protected endpoint
app.post('/submit', (req, res) => {
    res.send(`Success! Your message "${req.body.message}" was received securely.`);
});

// Route to simulate a CSRF attack by sending a request without the token
app.get('/attack', (req, res) => {
    res.send('Clicking this link simulates a GET request, but if you were to trigger a POST to /submit from an external site without the token, it would fail.');
});

// Example of a PUT request validation via API/Header
app.put('/update', (req, res) => {
    res.send('Resource updated successfully!');
});

const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
});